modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java

```diff
@@ -615,11 +613,21 @@ private <R extends NetworkMessage> CompletableFuture<R> sendWithRetry(
         ThrottlingContextHolder peerThrottlingContextHolder = throttlingContextHolder.peerContextHolder(peer.consistentId());
 
         if (throttleOnOverload && peerThrottlingContextHolder.isOverloaded()) {
-            executor.schedule(
-                    () -> future.completeExceptionally(new GroupOverloadedException(groupId, peer)),
-                    100,
-                    TimeUnit.MILLISECONDS
-            );
+            if (!busyLock.enterBusy()) {
```

Comment:
Why do we take the lock only inside the `if`?

Reply:
Well, all other stuff does not depend on the service's internal state. So, I decided to reduce the code that should be executed under the lock.